File: app-android/src/main/java/com/ustadmobile/port/android/view/ClazzDetailEnrollStudentActivity.java
package com.ustadmobile.port.android.view;

import android.arch.lifecycle.LiveData;
import android.arch.paging.DataSource;
import android.arch.paging.LivePagedListBuilder;
import android.arch.paging.PagedList;
import android.os.Bundle;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.widget.Button;

import com.toughra.ustadmobile.R;
import com.ustadmobile.core.controller.ClazzDetailEnrollStudentPresenter;
import com.ustadmobile.core.db.UmProvider;
import com.ustadmobile.core.view.ClazzDetailEnrollStudentView;
import com.ustadmobile.lib.db.entities.PersonWithEnrollment;
import com.ustadmobile.port.android.util.UMAndroidUtil;

import ru.dimorinny.floatingtextbutton.FloatingTextButton;

import static com.ustadmobile.core.controller.ClazzListPresenter.ARG_CLAZZ_UID;

/**
 * Clazz detail Enroll Student - Enrollment activity.
 * Gets called when "Add Student" is clicked on both the current Clazz selected as well as
 * from the People Bottom Navigation.
 */
public class ClazzDetailEnrollStudentActivity extends UstadBaseActivity implements
        ClazzDetailEnrollStudentView {

    //Toolbar
    private Toolbar toolbar;

    //Recycler View
    private RecyclerView mRecyclerView;
    private RecyclerView.LayoutManager mRecyclerLayoutManager;

    //Presenter
    private ClazzDetailEnrollStudentPresenter mPresenter;

    private long currentClazzUid;

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()){
            case android.R.id.home:
                onBackPressed();
                return true;
        }
        return super.onOptionsItemSelected(item);
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        //Set Layout
        setContentView(R.layout.activity_clazz_detail_enroll_student);

        //Toolbar:
        toolbar = findViewById(R.id.activity_clazz_detail_enroll_student_toolbar);
        toolbar.setTitle(getText(R.string.add_student));
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        //Get the clazz Uid from the arguments
        if(getIntent().hasExtra(ARG_CLAZZ_UID)){
            currentClazzUid = getIntent().getLongExtra(ARG_CLAZZ_UID, -1L);
        }

        //RecyclerView:
        mRecyclerView = findViewById(
                R.id.activity_clazz_detail_enroll_student_recycler_view);
        mRecyclerLayoutManager = new LinearLayoutManager(getApplicationContext());
        mRecyclerView.setLayoutManager(mRecyclerLayoutManager);

        //Presenter
        mPresenter = new ClazzDetailEnrollStudentPresenter(this,
                UMAndroidUtil.bundleToHashtable(getIntent().getExtras()), this);
        mPresenter.onCreate(UMAndroidUtil.bundleToHashtable(savedInstanceState));

        //Enroll new student
        Button newStudentButton = findViewById(R.id.activity_clazz_Detail_enroll_student_new);
        newStudentButton.setOnClickListener(v -> mPresenter.handleClickEnrollNewStudent());

        //FAB
        FloatingTextButton fab = findViewById(R.id.activity_clazz_detail_enroll_student_fab_done);
        fab.setOnClickListener(v -> mPresenter.handleClickDone());

        //TODO:
        //Filter / Sort
        //findViewById(R.id.activity_clazz_detail_enroll_student_filter)
        //        .setOnClickListener(mPresenter.handleChangeSortOrder(...));

        //Search:
        //findViewById(R.id.activity_clazz_detail_enroll_student_search)
        //        .setOnClickListener(mPresenter.handleClickSearch(...););

    }

    @Override
    public void setStudentsProvider(UmProvider<PersonWithEnrollment> studentsProvider) {

        PersonWithEnrollmentRecyclerAdapter recyclerAdapter =
                new PersonWithEnrollmentRecyclerAdapter(DIFF_CALLBACK, getApplicationContext(),
                        this, mPresenter, true, true);

        DataSource.Factory<Integer, PersonWithEnrollment> factory =
                (DataSource.Factory<Integer, PersonWithEnrollment>)
                        studentsProvider.getProvider();

        LiveData<PagedList<PersonWithEnrollment>> data =
                new LivePagedListBuilder<>(factory, 20).build();
        data.observe(this, recyclerAdapter::submitList);

        mRecyclerView.setAdapter(recyclerAdapter);
    }


    // Diff callback.
    public static final DiffUtil.ItemCallback<PersonWithEnrollment> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<PersonWithEnrollment>() {
                @Override
                public boolean areItemsTheSame(PersonWithEnrollment oldItem,
                                               PersonWithEnrollment newItem) {
                    return oldItem.getPersonUid() ==
                            newItem.getPersonUid();
                }

                @Override
                public boolean areContentsTheSame(PersonWithEnrollment oldItem,
                                                  PersonWithEnrollment newItem) {
                    return oldItem.equals(newItem);
                }
            };



}

File: app-android/src/main/kotlin/com/ustadmobile/port/android/view/ClazzDetailEnrollStudentActivity.kt
package com.ustadmobile.port.android.view

import android.os.Bundle
import android.view.MenuItem
import android.widget.Button
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Observer
import androidx.paging.DataSource
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.toughra.ustadmobile.R
import com.ustadmobile.core.controller.ClazzDetailEnrollStudentPresenter
import com.ustadmobile.core.impl.UMAndroidUtil
import com.ustadmobile.core.view.ClazzDetailEnrollStudentView
import com.ustadmobile.core.view.ClazzDetailEnrollStudentView.Companion.ARG_NEW_PERSON_TYPE
import com.ustadmobile.core.view.ClazzListView.Companion.ARG_CLAZZ_UID
import com.ustadmobile.core.view.GroupDetailView.Companion.GROUP_UID
import com.ustadmobile.lib.db.entities.ClazzMember
import com.ustadmobile.lib.db.entities.PersonWithEnrollment
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import ru.dimorinny.floatingtextbutton.FloatingTextButton

/**
 * Clazz detail Enroll Student - Enrollment activity.
 * Gets called when "Add Student" is clicked on both the current Clazz selected as well as
 * from the People Bottom Navigation.
 */
class ClazzDetailEnrollStudentActivity : UstadBaseActivity(), ClazzDetailEnrollStudentView {

    //Toolbar
    private var toolbar: Toolbar? = null

    //Recycler View
    private var mRecyclerView: RecyclerView? = null
    private var mRecyclerLayoutManager: RecyclerView.LayoutManager? = null

    //Presenter
    private var mPresenter: ClazzDetailEnrollStudentPresenter? = null

    private var currentClazzUid: Long = 0
    private var currentRole: Int = 0

    //PersonGroup enrollment
    private var groupUid: Long = 0

    private var groupEnrollment = false

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        //Set Layout
        setContentView(R.layout.activity_clazz_detail_enroll_student)

        //Toolbar:
        toolbar = findViewById(R.id.activity_clazz_detail_enroll_student_toolbar)

        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        //Get the clazz Uid from the arguments
        if (intent.hasExtra(ARG_CLAZZ_UID)) {
            currentClazzUid = intent.extras!!.get(ARG_CLAZZ_UID).toString().toLong()
        }

        if (intent.hasExtra(ARG_NEW_PERSON_TYPE)) {
            currentRole = intent.extras!!.get(ARG_NEW_PERSON_TYPE).toString().toInt()
        }

        //PersonGroup enrollment
        if (intent.hasExtra(GROUP_UID)) {
            groupUid = intent.extras!!.get(GROUP_UID).toString().toLong()
            if (groupUid != 0L) {
                groupEnrollment = true
            }
        }

        val enrollNewClazzMemberButton = findViewById<Button>(R.id.activity_clazz_Detail_enroll_student_new)

        if (currentRole == ClazzMember.ROLE_TEACHER) {
            toolbar!!.title = getText(R.string.add_teacher)
            enrollNewClazzMemberButton.text = getText(R.string.enroll_new_teacher)
        } else if (currentRole == ClazzMember.ROLE_STUDENT) {
            toolbar!!.title = getText(R.string.add_student)
            enrollNewClazzMemberButton.text = getText(R.string.enroll_new_student)
        } else if (groupUid != 0L) {
            //Add person to group
            toolbar!!.setTitle(R.string.enroll_group_member)
            enrollNewClazzMemberButton.text = getText(R.string.enroll_new_group_member)
        }

        //RecyclerView:
        mRecyclerView = findViewById(
                R.id.activity_clazz_detail_enroll_student_recycler_view)
        mRecyclerLayoutManager = LinearLayoutManager(applicationContext)
        mRecyclerView!!.layoutManager = mRecyclerLayoutManager

        //Presenter
        mPresenter = ClazzDetailEnrollStudentPresenter(this,
                UMAndroidUtil.bundleToMap(intent.extras), this)
        mPresenter!!.onCreate(UMAndroidUtil.bundleToMap(savedInstanceState))

        //Enroll new student
        val newStudentButton = findViewById<Button>(R.id.activity_clazz_Detail_enroll_student_new)
        newStudentButton.setOnClickListener { v -> mPresenter!!.handleClickEnrollNewPerson() }

        //FAB
        val fab = findViewById<FloatingTextButton>(R.id.activity_clazz_detail_enroll_student_fab_done)
        fab.setOnClickListener { v -> mPresenter!!.handleClickDone() }

    }

    override fun setStudentsProvider(factory: DataSource.Factory<Int, PersonWithEnrollment>) {

        val recyclerAdapter = PersonWithEnrollmentRecyclerAdapter(DIFF_CALLBACK, applicationContext,
                this, mPresenter!!, true, true, groupEnrollment)

        val data = LivePagedListBuilder(factory, 20).build()
        //Observe the data:
        val thisP = this
        GlobalScope.launch(Dispatchers.Main) {
            data.observe(thisP,
                    Observer<PagedList<PersonWithEnrollment>> { recyclerAdapter.submitList(it) })
        }

        mRecyclerView!!.adapter = recyclerAdapter
    }

    companion object {


        // Diff callback.
        val DIFF_CALLBACK: DiffUtil.ItemCallback<PersonWithEnrollment> = object : DiffUtil.ItemCallback<PersonWithEnrollment>() {
            override fun areItemsTheSame(oldItem: PersonWithEnrollment,
                                         newItem: PersonWithEnrollment): Boolean {
                return oldItem.personUid == newItem.personUid
            }

            override fun areContentsTheSame(oldItem: PersonWithEnrollment,
                                            newItem: PersonWithEnrollment): Boolean {
                return oldItem == newItem
            }
        }
    }


}
